dpos/state/checkpoint_test.go

```diff
@@ -0,0 +1,339 @@
+	result := &StateKeyFrame{
+		NodeOwnerKeys:             make(map[string]string),
+		PendingProducers:          make(map[string]*Producer),
+		ActivityProducers:         make(map[string]*Producer),
+		InactiveProducers:         make(map[string]*Producer),
+		CanceledProducers:         make(map[string]*Producer),
+		IllegalProducers:          make(map[string]*Producer),
+		Votes:                     make(map[string]*types.Output),
+		Nicknames:                 make(map[string]struct{}),
+		SpecialTxHashes:           make(map[common.Uint256]struct{}),
+		PreBlockArbiters:          make(map[string]struct{}),
+		EmergencyInactiveArbiters: make(map[string]struct{}),
+		VersionStartHeight:        rand.Uint32(),
+		VersionEndHeight:          rand.Uint32(),
+	}
+
+	for i := 0; i < 5; i++ {
+		result.NodeOwnerKeys[randomString()] = randomString()
+		result.PendingProducers[randomString()] = randomProducer()
+		result.ActivityProducers[randomString()] = randomProducer()
+		result.InactiveProducers[randomString()] = randomProducer()
+		result.CanceledProducers[randomString()] = randomProducer()
+		result.IllegalProducers[randomString()] = randomProducer()
+		result.Votes[randomString()] = randomVotes()
+		result.Nicknames[randomString()] = struct{}{}
+		result.SpecialTxHashes[*randomHash()] = struct{}{}
+		result.PreBlockArbiters[randomString()] = struct{}{}
+		result.EmergencyInactiveArbiters[randomString()] = struct{}{}
+	}
+	return result
+}
+
+func producerEqual(first *Producer, second *Producer) bool {
+	if first.state != second.state ||
+		first.registerHeight != second.registerHeight ||
+		first.cancelHeight != second.cancelHeight ||
+		first.inactiveCountingHeight != second.inactiveCountingHeight ||
+		first.inactiveSince != second.inactiveSince ||
+		first.activateRequestHeight != second.activateRequestHeight ||
+		first.illegalHeight != second.illegalHeight ||
+		first.penalty != second.penalty ||
+		first.votes != second.votes {
+		return false
+	}
+
+	return producerInfoEqual(&first.info, &second.info)
+}
+
+func producerInfoEqual(first *payload.ProducerInfo,
+	second *payload.ProducerInfo) bool {
+	if first.NickName != second.NickName ||
+		first.Url != second.Url ||
+		first.Location != second.Location ||
+		first.NetAddress != second.NetAddress {
+		return false
+	}
+
+	return bytes.Equal(first.OwnerPublicKey, second.OwnerPublicKey) &&
+		bytes.Equal(first.NodePublicKey, second.NodePublicKey) &&
+		bytes.Equal(first.Signature, second.Signature)
+}
+
+func rewardEqual(first *RewardData, second *RewardData) bool {
+	if first.TotalVotesInRound != second.TotalVotesInRound {
+		return false
+	}
+
+	if !hashesEqual(first.OwnerProgramHashes, second.OwnerProgramHashes) ||
+		!hashesEqual(first.CandidateOwnerProgramHashes,
+			second.CandidateOwnerProgramHashes) {
+		return false
+	}
+
+	return votesMapEqual(first.OwnerVotesInRound, second.OwnerVotesInRound)
+}
+
+func randomRewardData() *RewardData {
+	result := &RewardData{
```

Comment:
Here can use a NewRewardData method instead of duplicate initializations.

Reply:
Fixed